node/src/components/block_validator/tests.rs

```diff
@@ -0,0 +1,282 @@
+        module_bytes: Bytes::new(),
+        args: RuntimeArgs::new(),
+    };
+    let dependencies = vec![];
+    let gas_price = 1;
+
+    Deploy::new(
+        timestamp,
+        ttl,
+        gas_price,
+        dependencies,
+        chain_name,
+        payment,
+        session,
+        &secret_key,
+    )
+}
+
+fn new_transfer(rng: &mut TestRng, timestamp: Timestamp, ttl: TimeDiff) -> Deploy {
+    let secret_key = SecretKey::random(rng);
+    let chain_name = "chain".to_string();
+    let payment = ExecutableDeployItem::ModuleBytes {
+        module_bytes: Bytes::new(),
+        args: runtime_args! { ARG_AMOUNT => 1 },
+    };
+    let session = ExecutableDeployItem::Transfer {
+        args: RuntimeArgs::new(),
+    };
+    let dependencies = vec![];
+    let gas_price = 1;
+
+    Deploy::new(
+        timestamp,
+        ttl,
+        gas_price,
+        dependencies,
+        chain_name,
+        payment,
+        session,
+        &secret_key,
+    )
+}
+
+/// Validates a block using a `BlockValidator` component, and returns the result.
+async fn validate_block(
+    rng: &mut TestRng,
+    timestamp: Timestamp,
+    deploys: Vec<Deploy>,
+    transfers: Vec<Deploy>,
+) -> bool {
+    // Assemble the block to be validated.
+    let deploy_hashes = deploys.iter().map(|deploy| *deploy.id()).collect_vec();
+    let transfer_hashes = transfers.iter().map(|deploy| *deploy.id()).collect_vec();
+    let proposed_block = new_proposed_block(timestamp, deploy_hashes, transfer_hashes);
+
+    // Create the reactor and component.
+    let reactor = MockReactor::new();
+    let effect_builder = EffectBuilder::new(EventQueueHandle::new(reactor.scheduler));
+    let chainspec = Arc::new(Chainspec::from_resources("local"));
+    let mut block_validator = BlockValidator::<NodeId>::new(chainspec);
+
+    // Pass the block to the component. This future will eventually resolve to the result, i.e.
+    // whether the block is valid or not.
+    let validation_result =
+        tokio::spawn(effect_builder.validate_proposed_block("Bob", proposed_block.clone()));
+    let event = reactor.expect_block_validator_event().await;
+    let effects = block_validator.handle_event(effect_builder, rng, event);
+
+    // If validity could already be determined, the effect will be the validation response.
+    if block_validator.validation_states.is_empty() {
+        assert_eq!(1, effects.len());
+        for effect in effects {
+            tokio::spawn(effect).await.unwrap(); // Response.
+        }
+        return validation_result.await.unwrap();
+    }
+
+    // Otherwise the effects must be requests to fetch the block's deploys.
+    let fetch_results: Vec<_> = effects.into_iter().map(tokio::spawn).collect();
+
+    // We make our mock reactor answer with the expected deploys and transfers:
+    for deploy in deploys.into_iter().chain(transfers) {
+        reactor.expect_fetch_deploy(deploy).await;
+    }
+
```

Comment:
Note: Reactor should be empty now

Reply:
As discussed, I wouldn't explicitly check that here, to keep the test as simple as possible.